### PR TITLE
Add per-batch cuda stream completion wait

### PR DESCRIPTION
**Pull request by abramov-oleg**

### CODE REVIEW

File: savant/deepstream/pyfunc.py [hunk 1]
```diff
@@ -21,6 +22,7 @@ class NvDsPyFuncPlugin(BasePyFuncPlugin):
     def __init__(self, **kwargs):
         super().__init__(**kwargs)
         self._sources = SourceInfoRegistry()
+        self.frame_streams = []
```
Comment (denisvmedyantsev) [inline savant/deepstream/pyfunc.py:25]: I'm not sure we need cv2.cuda.Stream in any pyfunc. May be better to have it only in bgremover for now? @bwsw

**Review by denisvmedyantsev — CHANGES_REQUESTED** [review]

File: savant/deepstream/pyfunc.py (hunk 1 above)
Comment (bwsw) [reply savant/deepstream/pyfunc.py:25]: @denisvmedyantsev 100% disagree. Why the user would not want to handle `GpuMat` in arbitrary PyFunc?

File: docker/Dockerfile.deepstream [hunk 2]
```diff
@@ -112,7 +94,13 @@ COPY requirements/deepstream.txt requirements.txt
 RUN python -m pip install --no-cache-dir -r requirements.txt \
     && rm requirements.txt
 
-RUN python -m pip install "https://github.com/insight-platform/Savant/releases/download/v$SAVANT_VERSION/savant-$SAVANT_VERSION-py3-none-any.whl"
+COPY --from=savant_build /tmp/build/dist /libs/savant/dist/
+RUN python -m pip install --no-cache-dir /libs/savant/dist/*.whl
+
+RUN curl -fsS https://eu-central-1.linodeobjects.com/savant-data/opencv-packages/x86/OpenCV-4.6.0-x86_64-{dev,libs,licenses,python,scripts}.deb \
```
[the remark this answers is not in the document]
Comment (abramov-oleg) [reply docker/Dockerfile.deepstream]: Added

File: Makefile [hunk 3]
```diff
@@ -59,6 +59,20 @@ build-docs:
...
 	-t savant-docs:$(SAVANT_VERSION) .
 
+build-opencv: opencv-build-image opencv-copy-to-host
+
+opencv-build-image:
+	DOCKER_BUILDKIT=1 docker build \
+	--build-arg SAVANT_VERSION=$(SAVANT_VERSION) \
+	--build-arg DEEPSTREAM_VERSION=$(DEEPSTREAM_VERSION) \
+	-f docker/Dockerfile.deepstream-opencv$(PLATFORM_SUFFIX) \
+	-t savant-ds-opencv$(PLATFORM_SUFFIX) \
```
[the remark this answers is not in the document]
Comment (abramov-oleg) [reply Makefile:68]: Removed the Savant version from opencv image

File: Makefile [hunk 4]
```diff
@@ -59,6 +59,20 @@ build-docs:
 	-f docker/$(DOCKER_FILE) \
 	-t savant-docs:$(SAVANT_VERSION) .
 
+build-opencv: opencv-build-image opencv-copy-to-host
```
[the remark this answers is not in the document]
Comment (abramov-oleg) [reply Makefile:62]: Wrote a PREPARE_RELEASE.md doc.

File: docker/Dockerfile.deepstream-opencv [hunk 5]
```diff
@@ -0,0 +1,56 @@
+ARG DEEPSTREAM_VERSION
+ARG SAVANT_VERSION
+FROM nvcr.io/nvidia/deepstream:$DEEPSTREAM_VERSION-devel
+
+# Newer cmake ver. needed for "FindCUDAToolkit"
```
[the remark this answers is not in the document]
Comment (abramov-oleg) [reply docker/Dockerfile.deepstream-opencv]: Removed cmake upgrades.

python3-dev still needed to build python bindings.

File: docker/Dockerfile.deepstream [hunk 6]
```diff
@@ -1,5 +1,23 @@
 ARG SAVANT_VERSION
 ARG DEEPSTREAM_VERSION
+
+FROM ubuntu:20.04 as savant_build
```
[the remark this answers is not in the document]
Comment (abramov-oleg) [reply docker/Dockerfile.deepstream]: Removed ubuntu image usage.

File: samples/opencv_cuda_bg_remover_mog2/bgremover.py [hunk 7]
```diff
@@ -35,8 +35,9 @@ def process_frame(self, buffer: Gst.Buffer, frame_meta: NvDsFrameMeta):
         :param buffer: Gstreamer buffer with this frame's data.
         :param frame_meta: This frame's metadata.
         """
+        stream = self.get_cuda_stream()
```
Comment (dorgun) [inline samples/opencv_cuda_bg_remover_mog2/bgremover.py:38]: Replace `stream=artist.stream` with `stream=stream` in the function gaussian_filter.apply, back_sub.apply, ref_frame.copyTo.
It looks more logical to me.

File: docker/Dockerfile.deepstream-opencv [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        -D CPACK_BINARY_TGZ=OFF \
+        -D CPACK_BINARY_TZ=OFF \
+        /opencv/opencv
+
+RUN make -j$(nproc)
+RUN make -j$(nproc) install
+RUN make -j$(nproc) package \
+    && for file in *.deb; do mv "$file" "${file%.deb}_${DEEPSTREAM_VERSION}.deb"; done
+
+ENTRYPOINT cp OpenCV*.deb /out
```
Comment (dorgun) [inline docker/Dockerfile.deepstream-opencv:57]: It seems strange that we create a docker file only to build a package in it and then copy it to /out folder when we run the image by ENTRYPOINT.  This image is not used anywhere else. This is probably additional optimization, but the whole compilation could be implemented as a bash script and executed in docker image, made stage in gitflow with automatic copying to s3, if the versions suddenly changed. I suggest creating a task for this, so as not to delay the release. @denisvmedyantsev @bwsw

**Review by dorgun — CHANGES_REQUESTED** [review]

File: docker/Dockerfile.deepstream-opencv (hunk 8 above)
Comment (bwsw) [reply docker/Dockerfile.deepstream-opencv:57]: @dorgun I would create a ticket, but ignore that, for now, to not delay the release.

File: savant/deepstream/pyfunc.py [hunk 9]
```diff
@@ -34,6 +36,14 @@ def on_source_eos(self, source_id: str):
         """On source EOS event callback."""
         # self.logger.debug('Got GST_NVEVENT_STREAM_EOS for source %s.', source_id)
 
+    def get_cuda_stream(self):
```
[the remark this answers is not in the document]
Comment (abramov-oleg) [reply savant/deepstream/pyfunc.py]: Fixed.

File: samples/opencv_cuda_bg_remover_mog2/bgremover.py (hunk 7 above)
Comment (abramov-oleg) [reply samples/opencv_cuda_bg_remover_mog2/bgremover.py]: Fixed.

**Review by dorgun — APPROVED** [review]

**Review by denisvmedyantsev — CHANGES_REQUESTED** [review]

**Review by denisvmedyantsev — APPROVED** [review]

File: savant/deepstream/pyfunc.py [hunk 10]
```diff
@@ -40,7 +40,7 @@ def get_cuda_stream(self, frame_meta: NvDsFrameMeta):
         """Get a CUDA stream that can be used to asynchronously process a frame in a batch.
         All frame CUDA streams will be waited for at the end of batch processing.
         """
-        if frame_meta.batch_id not in self.frame_streams[frame_meta.batch_id]:
+        if frame_meta.batch_id not in self.frame_streams:
```
Comment (bwsw) [inline savant/deepstream/pyfunc.py:43]: Please add debug batch_id to ensure it works properly on batched operations.

File: PREPARE_RELEASE.md [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+This is required if the release includes an update to OpenCV or Deepstream versions, or to the savant OpenCV module (code in `libs/opencv/savant`).
+
+Run
+
+```
+make build-opencv
+```
+
+Do this for both `x86` and `aarch64` platforms.
```
[the remark this answers is not in the document]
Comment (abramov-oleg) [reply PREPARE_RELEASE.md]: Added docker buildx build for arm64.

File: savant/deepstream/pyfunc.py (hunk 10 above)
Comment (abramov-oleg) [reply savant/deepstream/pyfunc.py:47]: Added debug logging.

Tested `opencv_cuda_bg_remover_mog2` sample on 2 sources with batch_size=1 and batch_size=2.

**Pull request merged**